doc/rm/asm_coding_style.md

```diff
@@ -0,0 +1,279 @@
+---
+title: "RISC-V Assembly Style Guide"
+---
+
+## Basics
+
+### Summary
+
+OpenTitan needs to implement substantial functionality directly in RISC-V assembly.
+This document describes best practices for both assembly `.S` files and inline assembly statements in C and C++.
+It also codifies otherwise unwritten style guidelines in one central location.
+
+This document is not an introduction to RISC-V assembly; for that purpose, see the [RISC-V Assembly Programmer's Manual](https://github.com/riscv/riscv-asm-manual/blob/master/riscv-asm.md).
+
+Assembly is typically very specialized; the following rules do not presume to describe every use-case, so use your best judgement.
+
+This style guide is specialized for R32IMC, the ISA implemented by Ibex. 
+As such, no advice is provided for other RISC-V extensions, though this style guide is written such that advice for other extensions could be added without conflicts.
+
+## General Advice
+
+### Register Names
+
+***When referring to a RISC-V register, they must be referred to by their ABI names.***
+
+See the [psABI
+Reference](https://github.com/riscv/riscv-elf-psabi-doc/blob/master/riscv-elf.md#register-convention)
+for a reference to these names.
+
+Example:
+```S
+  // Correct:
+  li a0, 42
+  // Wrong:
+  li x10, 42
+```
+
+This rule can be ignored when the ABI meaning of a register is unimportant, e.g., such as when clobbering all 31 general-purpose registers.
+
+### Pseudoinstrucions
+
+***When performing an operation for which a pseudoinstruction exists, that pseudoinstruction must be used.***
+
+Pseudoinstructions make RISC-V's otherwise verbose RISC style more readable; for consistency, these must be used where possible.
+
+Example:
+```S
+  // Correct:
+  sw t0, _my_global, t1
+  // Wrong:
+  la t1, _my_global
+  sw t0, 0(t1)
+
+  // Correct:
+  ret
+  // Wrong:
+  jr ra
+```
+
+### Loading Addresses
+
+***Always use `la` to load the address of a symbol; always use `li` to load an address stored in a `#define`.***
+
+Some assemblers allow `la` with an immediate expression instead of a symbol, allowing a form of symbol+offset.
+However, support for this behavior is patchy, and the semantics of PIC `la` with immediate are unclear (in PIC mode, `la` should perform a GOT lookup, not a `pc`-relative load).
+
+### Jumping into C
+
+***Jumping into a C function must be done either with a `call` instruction, or, if that function is marked `noreturn`, a `tail` instruction.***
+
+Doing an unconditional jump with `j` into a C function can lead to erratic behavior if the C compiler emits a `ret` instruction for it.
+Furthermore, calling across translation units should be performed with a far jump, like `call` or `tail`, to avoid linking woes.
```

Comment:
Can we please be a bit more explicit here about what the problem is when you use the `j` and how `call` and `tail` solves it? Perhaps giving an example of how when you change how things are linked the `j` approach can break. Also, the meaning of a "far jump" might not be obvious to the reader, so consider something more explicit (e.g. "a pseudo-instruction that expands to the correct sequence", etc.)

Reply:
I've revised that paragraph. Let me know what you think.